plugins/tar/functions

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
+set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
+
+tar_build_cmd() {
+  declare desc="calls tar:build-locked via command line"
+  local cmd="tar:build"
+
+  verify_app_name "$2"
+  local APP="$2"; local APP_BUILD_LOCK="$DOKKU_ROOT/$APP/.build.lock"
+  local APP_BUILD_LOCK_MSG="$APP is currently being deployed or locked. Waiting..."
+  [[ $(flock -n "$APP_BUILD_LOCK" true &>/dev/null ; echo $?) -ne 0 ]] && echo "$APP_BUILD_LOCK_MSG"
+
+  shift 1
+  flock -o "$APP_BUILD_LOCK" tar_build_locked_cmd "tar:build-locked" "$@"
+}
+
+tar_build_locked_cmd() {
+  declare desc="builds apps from tarball via command line"
+  local cmd="tar:build-locked"
+  local APP="$2"
+
+  verify_app_name "$2"
+  shift 2
+
+  # clean up after ourselves
+  local TAR_BUILD_TMP_WORK_DIR=$(mktemp -d "/tmp/dokku_tar.XXXX")
+  trap 'rm -rf "$TAR_BUILD_TMP_WORK_DIR" > /dev/null' RETURN INT TERM EXIT
+
+  # extract tar file
+  chmod 755 "$TAR_BUILD_TMP_WORK_DIR"
+  pushd "$TAR_BUILD_TMP_WORK_DIR" > /dev/null
+
+  # Detect a common prefix that all files in the tar have, and strip off each directory found in it
+  local COMMON_PREFIX=$(tar -tf "$DOKKU_ROOT/$APP/src.tar" | sed -e 'N;s/^\(.*\).*\n\1.*$/\1\n\1/;D')
+  local BOGUS_PARTS=$(echo "$COMMON_PREFIX " | awk 'BEGIN{FS="/"} {print NF-1}')
+
+  dokku_log_info1_quiet "Striping $BOGUS_PARTS worth of directories from tarball"
+
+  tar -x -C "$TAR_BUILD_TMP_WORK_DIR" -f "$DOKKU_ROOT/$APP/src.tar" --strip-components="$BOGUS_PARTS"
+  chmod -R u+r "$TAR_BUILD_TMP_WORK_DIR"
+
+  if [[ -f Dockerfile ]] && [[ "$([[ -f .env ]] && grep -q BUILDPACK_URL .env; echo $?)" != "0" ]] && [[ ! -f ".buildpacks" ]]; then
+    plugn trigger pre-receive-app "$APP" "dockerfile" "$TAR_BUILD_TMP_WORK_DIR"
+    dokku receive "$APP" "dockerfile" "$TAR_BUILD_TMP_WORK_DIR" | sed -u "s/^/"$'\e[1G'"/"
+  else
+    plugn trigger pre-receive-app "$APP" "herokuish" "$TAR_BUILD_TMP_WORK_DIR"
+    dokku receive "$APP" "herokuish" "$TAR_BUILD_TMP_WORK_DIR" | sed -u "s/^/"$'\e[1G'"/"
```

Comment:
Understood. For the record, you don't need to PR to make tests run. CI will run branch pushed to the project repo.